repo/content/index_blob_manager.go

```diff
@@ -0,0 +1,456 @@
+}
+
+func (m *indexBlobManagerImpl) getCompactionLogEntries(ctx context.Context, blobs []blob.Metadata) (map[blob.ID]*compactionLogEntry, error) {
+	results := map[blob.ID]*compactionLogEntry{}
+
+	for _, cb := range blobs {
+		data, err := m.getEncryptedBlob(ctx, cb.BlobID)
+
+		if errors.Is(err, blob.ErrBlobNotFound) {
+			continue
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to read compaction blob %q", cb.BlobID)
+		}
+
+		le := &compactionLogEntry{}
+
+		if err := json.Unmarshal(data, le); err != nil {
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (m *indexBlobManagerImpl) getCleanupEntries(ctx context.Context, latestServerBlobTime time.Time, blobs []blob.Metadata) (map[blob.ID]*cleanupEntry, error) {
+	results := map[blob.ID]*cleanupEntry{}
+
+	for _, cb := range blobs {
+		data, err := m.getEncryptedBlob(ctx, cb.BlobID)
+
+		if errors.Is(err, blob.ErrBlobNotFound) {
+			continue
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to read compaction blob %q", cb.BlobID)
+		}
+
+		le := &cleanupEntry{}
+
+		if err := json.Unmarshal(data, le); err != nil {
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		le.age = latestServerBlobTime.Sub(cb.Timestamp)
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (m *indexBlobManagerImpl) deleteOldBlobs(ctx context.Context, latestBlob blob.Metadata) error {
+	allCompactionLogBlobs, err := m.listCache.listBlobs(ctx, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing compaction log blobs")
+	}
+
+	// look for server-assigned timestamp of the compaction log entry we just wrote as a reference.
+	// we're assuming server-generated timestamps are somewhat reasonable and time is moving
+	compactionLogServerTimeCutoff := latestBlob.Timestamp.Add(-m.maxEventualConsistencySettleTime)
+	compactionBlobs := blobsOlderThan(allCompactionLogBlobs, compactionLogServerTimeCutoff)
+
+	log(ctx).Debugf("fetching %v/%v compaction logs older than %v", len(compactionBlobs), len(allCompactionLogBlobs), compactionLogServerTimeCutoff)
+
+	compactionBlobEntries, err := m.getCompactionLogEntries(ctx, compactionBlobs)
+	if err != nil {
+		return errors.Wrap(err, "unable to get compaction log entries")
+	}
+
+	allCleanupBlobs, err := m.listCache.listBlobs(ctx, cleanupBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing cleanup blobs")
+	}
+
+	cleanupEntries, err := m.getCleanupEntries(ctx, latestBlob.Timestamp, allCleanupBlobs)
+	if err != nil {
+		return errors.Wrap(err, "error loading cleanup blobs")
+	}
+
+	indexBlobsToDelete := m.findIndexBlobsToDelete(ctx, latestBlob.Timestamp, compactionBlobEntries)
+	compactionLogBlobsToDelayCleanup := m.findCompactionLogBlobsToDelayCleanup(ctx, compactionBlobs)
+
+	if err := m.delayCleanupBlobs(ctx, compactionLogBlobsToDelayCleanup); err != nil {
```

Comment:
There's an issue here with the order of creation of the cleanup blobs. I'll elaborate in a few.

Reply:
The creation of the delay cleanup blobs should only be done _after_ the inputs for a given compaction have been removed. At this point in time (`t_c`, then we are saying:
* All the inputs for this compaction have been deleted have been deleted prior to `t_c`, then the compaction log (and the cleanup log) can be safely deleted at `t > t_c + ec_settle_time`

Reply:
good catch, fixed